bin/varnishd/cache/cache_req.c

```diff
@@ -173,7 +173,7 @@ Req_New(struct sess *sp)
 	p = (void*)PRNDUP(p + sizeof(*req->vfc));
 
 	req->vdc = (void*)p;
-	memset(req->vdc, 0, sizeof *req->vdc);
+	INIT_OBJ(req->vdc, VDP_CTX_MAGIC);
```

Comment:
we only add a magic value in the respective struct's initializer, in this case `VDP_Init()` or such initializer does not exist as a function and initialization is basically _just_ the `INIT_OBJ()`. This is the case for the other structs initialized here.

Can we find a better way to make sanitizers happy?

Reply:
> Can we find a better way to make sanitizers happy?

I understand that mempools already back our main workspaces, but my suggestion would be to make the mempools have their own "mpl" workspace. This way we could allocate struct req, struct busyobj, and other things as workspace allocations as we do today, and allocate the task workspace from whatever remains from the mempool workspace.

This way, when working with sanitizers, the workspace emulator would turn those pointer-arithmetic pseudo-allocations we have today into individual mallocs and we'd finally be able to trigger an overflow in structs such as req and busyobj.

Reply:
helpful clarification from bugwash: What Dridi means is `req = WS_Alloc(sizeof *req)`

Reply:
Sorry I haven't had much time to get back to this. I'm not sure initially where we'd get the workspace from. The session? Since this function is the one that initializes the workspace itself.

Noted in here: https://github.com/varnishcache/varnish-cache/pull/4062#issuecomment-1971312126 I realized the issue is more from using (my version of) GCC with ubsan than my distro itself. So using `ZERO_OBJ` instead of an explicit memset passes compilation fine. Though this feels more like tricking the compiler but the compiler was not fully right in the first place.